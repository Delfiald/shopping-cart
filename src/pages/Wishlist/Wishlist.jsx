import { useOutletContext, useSearchParams } from "react-router-dom";
import WishlistMain from "../../components/Main/WishlistMain";
import { useEffect } from "react";

function Wishlist() {
 const [searchParams, setSearchParams] = useSearchParams();
 const {
  products,
  setCartItem,
  wishlistItem,
  setWishlistItem,
  hoverButton,
  setHoverButton,
  isExiting,
  setIsExiting,
 } = useOutletContext();

 const page = searchParams.get("page");
 const itemPerPage = searchParams.get("itemsPerPage");
 const sort = searchParams.get("sort");

 const handlePageChange = (page) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", page);
   return params;
  });
 };

 const handleItemsPerPageChange = (itemsPerPage) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", 1);
   params.set("itemsPerPage", itemsPerPage);

   return params;
  });
 };

 const handleSortChange = (sort) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("sort", sort);
   return params;
  });
 };

 const getWishlistDetails = () => {
  return wishlistItem.map((wishlist) => {
   const product = products.find((p) => p.id === wishlist.id);
   return {
    ...product,
   };
  });
 };

 useEffect(() => {
  const validSort = [
   "name-asc",
   "name-desc",
   "price-asc",
   "price-desc",
  ].includes(sort)
   ? sort
   : "name-asc";
  const totalItems = wishlistItem.length;
  const maxPage =
   itemPerPage === "Infinity" ? 1 : Math.ceil(totalItems / (itemPerPage || 10));
  const validPage = Math.max(1, Math.min(parseInt(page) || 1, maxPage));
  const validItemsPerPage =
   itemPerPage === "Infinity"
    ? Infinity
    : [5, 10].includes(parseInt(itemPerPage))
    ? itemPerPage
    : 10;

  const defaultParams = {
   sort: validSort,
   page: validPage,
   itemsPerPage: validItemsPerPage,
  };

  const updatedParams = { ...defaultParams };

  if (
   JSON.stringify(updatedParams) !==
   JSON.stringify(Object.fromEntries(searchParams))
  ) {
   setSearchParams(updatedParams, { replace: true });
  }
 }, [
  page,
  itemPerPage,
  searchParams,
  setSearchParams,
  wishlistItem.length,
  sort,
 ]);

 return (
  <>
   <WishlistMain
    products={getWishlistDetails()}
    setCartItem={setCartItem}
    wishlistItem={wishlistItem}
    setWishlistItem={setWishlistItem}
    page={parseInt(page)}
    setPage={handlePageChange}
    itemPerPage={parseInt(itemPerPage)}
    setItemPerPage={handleItemsPerPageChange}
    sort={sort}
    setSort={handleSortChange}
    hoverButton={hoverButton}
    setHoverButton={setHoverButton}
    isExiting={isExiting}
    setIsExiting={setIsExiting}
   />
  </>
 );
}

export default Wishlist;
