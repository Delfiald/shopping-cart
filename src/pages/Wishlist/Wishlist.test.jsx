import { render } from "@testing-library/react";
import { MemoryRouter, Outlet, Route, Routes } from "react-router-dom";
import Wishlist from "./Wishlist";

const generateMockProducts = (num) => {
 const mockProducts = [];

 for (let i = 1; i <= num; i++) {
  mockProducts.push({
   id: i,
   title: `product-${i}`,
   image: `/image/product-${i}.png`,
   price: 100 + (i % 10) * 10,
   category: `category-${(i % 5) + 1}`,
  });
 }
 return mockProducts;
};

const generateMockWishlist = (num) => {
 const products = [];
 for (let i = 1; i <= num; i++) {
  products.push({
   id: i,
  });
 }
 return products;
};

const MockRouter = () => {
 return (
  <MemoryRouter initialEntries={["/wishlist"]}>
   <Routes>
    <Route
     path="/"
     element={
      <Outlet
       context={{
        products: generateMockProducts(3),
        wishlistItem: generateMockWishlist(2),
        setWishlistItem: vi.fn(),
       }}
      />
     }
    >
     <Route path="wishlist" element={<Wishlist />} />
    </Route>
   </Routes>
  </MemoryRouter>
 );
};

describe("Test Wishlist Page", () => {
 it("Should Render Wishlist Page", () => {
  const { container } = render(<MockRouter />);

  expect(container).toMatchSnapshot();
 });
});
