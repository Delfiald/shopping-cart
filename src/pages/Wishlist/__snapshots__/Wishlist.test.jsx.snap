// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Wishlist Page > Should Render Wishlist Page 1`] = `
<div>
  <main
    class="_wishlist_716e15 _fade-in_716e15 "
  >
    <h2>
      Wishlist
    </h2>
    <div
      class="_product-list-header_716e15"
    >
      <div
        data-testid="list-header"
      >
        <div
          class="_item-information_716e15"
        >
          Showing 1 - 2 products
        </div>
        <div
          class="_option-wrapper_716e15"
        >
          <div
            class="_sort-button_716e15"
            data-testid="sort-button"
          >
            <div
              data-testid="sort-value"
            >
              <p>
                Name Asc
              </p>
              <svg
                class="lucide lucide-chevron-down"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
          <div
            class="_item-per-page_716e15"
            data-testid="item-per-page"
          >
            <div
              class="_item-per-page-value_716e15"
              data-testid="item-per-page-value"
            >
              <p>
                10
              </p>
              <svg
                class="lucide lucide-chevron-down"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <section
      class="_product-list-wrapper_716e15"
      data-testid="product-card"
    >
      <div
        class="_card_716e15"
      >
        <div
          data-testid="product-card-1"
        >
          <div
            class="_image-wrapper_716e15"
          >
            <img
              alt="product-1"
              src="/image/product-1.png"
            />
          </div>
          <div
            class="_product-name_716e15"
          >
            product-1
          </div>
          <div
            class="_product-price_716e15"
          >
            $110
          </div>
        </div>
        <div
          class="_card-action_716e15"
        >
          <button
            class="_wishlist-button_716e15 _active_716e15"
            data-testid="wishlist-button-1"
          >
            <svg
              class="lucide lucide-heart"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
              />
            </svg>
          </button>
          <button
            class="_add-to-cart_716e15"
            data-testid="add-to-cart-button-1"
          >
            <div
              class="_displayed_716e15"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
              <p>
                Cart
              </p>
            </div>
            <div
              class="_hovered_716e15"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
              <p>
                Cart
              </p>
            </div>
          </button>
        </div>
      </div>
      <div
        class="_card_716e15"
      >
        <div
          data-testid="product-card-2"
        >
          <div
            class="_image-wrapper_716e15"
          >
            <img
              alt="product-2"
              src="/image/product-2.png"
            />
          </div>
          <div
            class="_product-name_716e15"
          >
            product-2
          </div>
          <div
            class="_product-price_716e15"
          >
            $120
          </div>
        </div>
        <div
          class="_card-action_716e15"
        >
          <button
            class="_wishlist-button_716e15 _active_716e15"
            data-testid="wishlist-button-2"
          >
            <svg
              class="lucide lucide-heart"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
              />
            </svg>
          </button>
          <button
            class="_add-to-cart_716e15"
            data-testid="add-to-cart-button-2"
          >
            <div
              class="_displayed_716e15"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
              <p>
                Cart
              </p>
            </div>
            <div
              class="_hovered_716e15"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
              <p>
                Cart
              </p>
            </div>
          </button>
        </div>
      </div>
    </section>
    <div
      class="_product-list-bottom_716e15"
    >
      <div
        class="_pagination_716e15"
      >
        <div
          class="_arrow-button_716e15"
        />
        <div
          class="_number-button_716e15"
        >
          <div
            class="_active_716e15"
            data-testid="current-page-button"
          >
            1
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
